Add a notebook for EBL correction example

AtreyeeS opened the pull request:
This PR adds a notebook to show how to correct for EBL absorption. 
The main idea is to show how to correctly set the covariance as discussed in #5218 

I decided to add a new notebook because it gets difficult to find otherwise. I am trying to see if, for this specific case, we can give a way to get the correct covariance for a free alpha/redshift

bkhelifi commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,232 @@
+# Model the observed spectrum
+# ---------------------------
+#
+# The observed spectrum is already attenuated due to the EBL. Assuming
```
```suggestion
# The observed spectrum is attenuated due to the EBL. Assuming
```

bkhelifi commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,232 @@
+
+
+######################################################################
+# We keep the paramters of the EBL model frozen in this example. For a
```
```suggestion
# We keep the parameters of the EBL model frozen in this example. For a
```

bkhelifi commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,232 @@
+fit = Fit()
+result = fit.run(datasets=[stacked])
+
+# we make a copy here to compare it later
```
```suggestion
# We make a copy here to compare it later.
```

bkhelifi commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,232 @@
+
+
+######################################################################
+# Get the flux points
```
```suggestion
# Get the flux points.
```

bkhelifi commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,232 @@
+# ===================
+#
+# To get the observed flux points, just run the `FluxPointsEstimator`
+# normally
```
```suggestion
# normally.
```

bkhelifi commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,232 @@
+######################################################################
+# To get the deabsorbed flux points (ie, intrinsic points), we simply need
+# to set the reference model to the best fit power law instead of the
+# compound model. We first make a copy of the computed flux points
```
```suggestion
# compound model. We first make a copy of the computed flux points.
```

bkhelifi reviewed (dismissed):
Thanks a lot @AtreyeeS . Except few details, it looks ok for me

registerrier reviewed (commented):
Thanks @AtreyeeS . This is nice.
The main issue is general: most of the tutorial is data reduction which is not the actual scope. I would rather rely on pre-reduced datasets in GAMMAPY-DATA to focus on what is specific here.

AtreyeeS commented:
> The main issue is general: most of the tutorial is data reduction which is not the actual scope. I would rather rely on pre-reduced datasets in GAMMAPY-DATA to focus on what is specific here.

The pre-reduced datasets in GAMMAPY-DATA are all galactic sources (crab and the galactic centre). For the PKS 2155-304 flare, we have the light curve (DL5).
I initially thought of adding the spectral datasets used here in GAMMAPY-DATA, but then it would be used only in this tutorial so I was not sure if its worth it.

Astro-Kirsty commented:
+1 for adding the spectral dataset to GAMMAPY_DATA
It could also come in handy when we have examples within the code!

AtreyeeS commented:
I modified the notebook to use precomputed spectral datasets and also added an example for fitting the redshift

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+
+
+######################################################################
+# We add Gaussian priors on the spectral parameters assuming the 4FGL
+# measurements. For more details on using priors, see
```
```suggestion
# We add Gaussian priors on the `alpha` and `beta` parameters based on the 4FGL
# measurements and the associated errors. For more details on using priors, see
```

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+# compound model. We first make a copy of the computed flux points
+#
+
+flux_points_intrinsic = deepcopy(flux_points_obs)
```
I think this suggests we need a `copy` method on `FluxMaps` and their subclasses.

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+flux_points_intrinsic = deepcopy(flux_points_obs)
+flux_points_intrinsic._reference_model = SkyModel(spectral_model=pwl)
+
+print(flux_points_obs._reference_model)
```
```suggestion
print(flux_points_obs.reference_model)
```
Can't you directly call the property `reference_model` itself?

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+
+print(flux_points_obs._reference_model)
+
+print(flux_points_intrinsic._reference_model)
```
```suggestion
print(flux_points_intrinsic.reference_model)
```

QRemy replied:
no the setter is missing

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+#
+
+flux_points_intrinsic = deepcopy(flux_points_obs)
+flux_points_intrinsic._reference_model = SkyModel(spectral_model=pwl)
```
A `FluxMaps.copy(reference_model=SkyModel())` could be a way to avoid setting the internal property. And a warning could be raised as this can lead to weird results if not done properly.

QRemy replied:
Done in #5279

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+# Now, create a sky model and proceed with the fit
+sky_model = SkyModel(spatial_model=None, spectral_model=spectral_model, name="pks2155")
+
+dataset.models = sky_model
```
Maybe remind here as a comment or in the text cell, that since this dataset has been produced by a reflected region analysis, it uses ON-OFF statistic and does not require a background model.

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+print(flux_points_intrinsic._reference_model)
+
+
+######################################################################
```
I think this part should be solved as a fix to the covariance handling in `CompoundSpectralModel`. Users shouldn't have to do this.

registerrier reviewed (dismissed):
Thanks @AtreyeeS ! This is really nice. I have left a few comments inline. I have a few more general ones:
- The issue with covariance matrix should be fixed upstream and should not require the fix in the tutorial. This can be done in a future PR, but we need a reminder issue (maybe in #5218 ).
- We might need a `FluxMaps.copy()` method rather than running `deepcopy` directly. This can be done in a future PR as well.
- Fitting the redshift using the Fermi measurement is nice. I just wonder whether this is the most common practice vs e.g. fitting the `alpha` parameter. Any opinion here?

Astro-Kirsty commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+#
+
+
+######################################################################
```
This section in the docs is a little hard to read so I have suggested some changes to help with that. Specifically adding extra lines between the print statements

Astro-Kirsty commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+
+print(EBL_DATA_BUILTIN.keys())
+
+# define the power law
```
```suggestion
######################################################################
# Define the power law
#
```

Astro-Kirsty commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+spectral_model = pwl * absorption
+print(spectral_model)
+
+# Now, create a sky model and proceed with the fit
```
```suggestion
######################################################################
# Now, create a sky model and proceed with the fit
#
```

Astro-Kirsty commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+spectral_model.covariance.plot_correlation()
+plt.show()
+
+# The covariance matrix on the power law does not contain the off diagnoal terms
```
```suggestion
######################################################################
# The covariance matrix on the power law does not contain the off diagonal terms
#
```

Astro-Kirsty commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+pwl.covariance.plot_correlation()
+plt.show()
+
+# Extract the sub covaraince and set is on the `pwl`
```
```suggestion
######################################################################
# Extract the sub covariance and set is on the `pwl`
#
```

Astro-Kirsty reviewed (dismissed):
Thanks a lot for the effort @AtreyeeS 

I have made some comments regarding format after building the docs!

Astro-Kirsty commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,292 @@
+
+
+######################################################################
+# Set the covariance on the power law model correctly
```
Nice catch, yes it should be replaced with the '---'

morcuended commented:
> The issue with covariance matrix should be fixed upstream and should not require the fix in the tutorial. This can be done in a future PR, but we need a reminder issue (maybe in https://github.com/gammapy/gammapy/issues/5218 ).

@AtreyeeS I can take care of this part in case you or someone else haven't jumped into it yet

AtreyeeS commented:
Thanks for all your comments. I updated the notebook after the recent changes in the covariance handling.
I also added a reference to `ebltable` which has an interface to gammapy (https://github.com/me-manu/ebltable/blob/master/notebooks/gammapy-model.ipynb), as we sometimes get questions on how to use non built in ebl models

registerrier reviewed (approved):
Thanks @AtreyeeS . Looks good to me!

AtreyeeS commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,282 @@
+# To use other EBL models, you need to save the optical depth as a
+# function of energy and redshift as an XSPEC model.
+# Alternatively, you can use packages like `ebltable <https://github.com/me-manu/ebltable>`
+# which interface ebl models with Gammapy.
```
Yes, I thought their example seems over-complicated, but something is probably better than nothing

morcuended commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,282 @@
+plt.ylim(bottom=1e-13)
+plt.legend()
+plt.show()
+# sphinx_gallery_thumbnail_number = 2
```
is this line needed or just a leftover?

morcuended commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,282 @@
+######################################################################
+# To use other EBL models, you need to save the optical depth as a
+# function of energy and redshift as an XSPEC model.
+# Alternatively, you can use packages like `ebltable <https://github.com/me-manu/ebltable>`
```
when I build the docs this link does not work for me

registerrier replied:
```suggestion
# Alternatively, you can use packages like `ebltable <https://github.com/me-manu/ebltable>`_
```

morcuended commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,282 @@
+# To use other EBL models, you need to save the optical depth as a
+# function of energy and redshift as an XSPEC model.
+# Alternatively, you can use packages like `ebltable <https://github.com/me-manu/ebltable>`
+# which shows how to interface other ebl models with Gammapy.
```
```suggestion
# which shows how to interface other EBL models with Gammapy.
```

morcuended commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,282 @@
+# Get the flux points
+# ===================
+#
+# To get the observed flux points, just run the `~gammapy.estimators.FluxPointsEstimator
```
```suggestion
# To get the observed flux points, just run the `~gammapy.estimators.FluxPointsEstimator`
```

registerrier commented on examples/tutorials/analysis-1d/ebl.py:
```diff
@@ -0,0 +1,282 @@
+# As usual, we’ll start with the standard imports …
+#
+
+from copy import deepcopy
```
```suggestion
```

registerrier commented:
Let's wait for CI and merge.

registerrier reviewed (approved):
Thanks @AtreyeeS !
